sipyco/pc_rpc.py

```diff
@@ -296,126 +371,133 @@ class BestEffortClient:
+        # Let the thread complete I/O and then do the socket closing.
+        # Python fails to provide a way to cancel threads...
+        self.__conmonitor_terminate = True
 
-    def __send(self, obj):
-        line = pyon.encode(obj) + "\n"
-        self.__socket.sendall(line.encode())
-
-    def __recv(self):
-        buf = self.__socket.recv(4096).decode()
-        while "\n" not in buf:
-            more = self.__socket.recv(4096)
-            if not more:
-                break
-            buf += more.decode()
-        return pyon.decode(buf)
+    def is_closed(self):
+        if not self.__client:
+            return True
 
-    def __do_rpc(self, name, args, kwargs):
-        if self.__conretry_thread is not None:
-            return None
-
-        obj = {"action": "call", "name": name, "args": args, "kwargs": kwargs}
-        try:
-            self.__send(obj)
-            obj = self.__recv()
-        except:
-            logger.warning("connection failed while attempting "
-                           "RPC to %s:%d[%s], re-establishing connection "
-                           "in the background",
-                           self.__host, self.__port, self.__target_name)
-            self.__start_conretry()
-            return None
-        else:
-            if obj["status"] == "ok":
-                return obj["ret"]
-            elif obj["status"] == "failed":
-                raise_packed_exc(obj["exception"])
-            else:
-                raise ValueError
+        with self.__client_lock:
+            return self.__client.is_closed()
 
     def __getattr__(self, name):
-        if name not in self.__valid_methods:
-            raise AttributeError
-        def proxy(*args, **kwargs):
-            return self.__do_rpc(name, args, kwargs)
-        return proxy
+        """
+        If the client is connected, pass this call onto the client. If not,
+        return a method that returns None if the call was valid or if we can't
+        tell. Raise AttributeError otherwise.
+        """
 
-    def get_selected_target(self):
-        raise NotImplementedError
+        if (
+            self.__valid_methods
+            and name not in self.__valid_methods
+            and name not in dir(Client)
+        ):
+            raise AttributeError("{} not found in cached list".format(name))
+        else:
+            # If the client exists, pass the call to the client
+            def tryer(*args, **kwargs):
+                if not self.__client:
+                    return None
 
-    def get_local_host(self):
-        raise NotImplementedError
+                with self.__client_lock:
+                    client_method = getattr(self.__client, name)
+                    try:
+                        return client_method(*args, **kwargs)
+                    except ConnectionError:
+                        self.__mark_invalid()
+                        return None
+            return tryer
```

Comment:
There's a bug here: if the `Server` returns a `ConnectionError` as the valid, unpacked output of its method this will be misinterpreted as a failed connection. I've added a failing test to highlight this, will have a think as to how to solve.

Reply:
This is resolved in the latest commit: Client now raises `RPCConnectionError` which is a subclass of `ConnectionError` if an RPC connection error occurs, which allows you to tell the difference between errors arising from the RPC code and errors that were unpacked. This is backwards compatible because any it's still a `ConnectionError` so any code which looks for these will still receive them.